src/sink/kv.rs

```diff
@@ -0,0 +1,275 @@
+    ///    "old": "<value|null|not-se>",
+    ///    "new": "<value|null|not-set>",
+    /// }
+    /// ```
+    Cas {
+        key: Vec<u8>,
+        old: Option<&'v Value<'v>>,
+        new: Option<&'v Value<'v>>,
+    },
+}
+
+#[async_trait::async_trait]
+impl Sink for Kv {
+    async fn on_event(
+        &mut self,
+        _input: &str,
+        codec: &mut dyn Codec,
+        _codec_map: &HashMap<String, Box<dyn Codec>>,
+        event: Event,
+    ) -> ResultVec {
+        let mut r = Vec::with_capacity(10);
+        let ingest_ns = tremor_common::time::nanotime();
+
+        let cmds = event.value_iter().filter_map(|v| {
+            if let Some(g) = v.get("get") {
+                Some(Command::Get {
+                    key: g.get_bytes("key")?.to_vec(),
+                })
+            } else if let Some(p) = v.get("put") {
+                Some(Command::Put {
+                    key: p.get_bytes("key")?.to_vec(),
+                    value: p.get("value")?,
+                })
+            } else if let Some(c) = v.get("cas") {
+                Some(Command::Cas {
+                    key: c.get_bytes("key")?.to_vec(),
+                    old: c.get("old"),
+                    new: c.get("new"),
+                })
+            } else if let Some(d) = v.get("delete") {
+                Some(Command::Delete {
+                    key: d.get_bytes("key")?.to_vec(),
+                })
+            } else if let Some(s) = v.get("scan") {
+                Some(Command::Scan {
+                    start: s.get_bytes("start").map(|v| v.to_vec()),
+                    end: s.get_bytes("end").map(|v| v.to_vec()),
+                })
+            } else {
+                error!("failed to decode command: {}", v);
+                None
+            }
+        });
+
+        for c in cmds {
+            println!("command: {:?}", c);
+            let data = self.execute(c, codec, ingest_ns)?;
+            let e = Event {
+                data: data.into(),
+                origin_uri: Some(self.event_origin_uri.clone()),
+                ..Event::default()
+            };
+            r.push(Reply::Response(OUT, e))
+        }
+        Ok(Some(r))
+    }
+
+    async fn on_signal(&mut self, _signal: Event) -> ResultVec {
+        Ok(None)
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    async fn init(
+        &mut self,
+        sink_uid: u64,
+        sink_url: &TremorURL,
+        _codec: &dyn Codec,
+        _codec_map: &HashMap<String, Box<dyn Codec>>,
+        _processors: Processors<'_>,
+        _is_linked: bool,
+        _reply_channel: Sender<sink::Reply>,
+    ) -> Result<()> {
+        self.event_origin_uri.uid = sink_uid;
+        self.sink_url = sink_url.clone();
+        Ok(())
+    }
+
+    fn is_active(&self) -> bool {
+        true
+    }
+
+    fn auto_ack(&self) -> bool {
+        true
+    }
+
+    fn default_codec(&self) -> &str {
+        "json"
+    }
+
+    async fn terminate(&mut self) {}
```

Comment:
should we flush the sled db here?

Reply:
I'm 99% sure that sled flushes when it's dropped

Reply:
i couldnt find anything in the docs or sled sources. Any pointer for me?

Reply:
It's not spelled out but implied in the examples, also it is what the code does :D (hard to find :sob:) https://docs.rs/sled/0.34.6/src/sled/tree.rs.html#98

We should probably ask it to be mentioned explicitly in the sled-documentation to ensure that it is part of the API contract.